### PR TITLE
[Merged by Bors] - feat: dividing an element of a euclidean domain by its radical

### PR DESCRIPTION
**Pull request by seewoo5**
For `a` in a `EuclideanDomain`, define `divRadical a` as `a / radical a`

---

Prove some related theorems.
This is needed in the proof of Mason-Stothers theorem, and also a part of PR #15706.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Radical.lean [hunk 1]
```diff
@@ -168,3 +179,51 @@ theorem radical_neg {a : R} : radical (-a) = radical a :=
   radical_eq_of_associated Associated.rfl.neg_left
 
 end UniqueFactorizationDomain
+
+open UniqueFactorizationDomain
+namespace EuclideanDomain
+
+variable {E : Type*} [EuclideanDomain E] [NormalizationMonoid E] [UniqueFactorizationMonoid E]
+
+/-- For an element `a` in an Euclidean domain, `a / radical a`. -/
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Radical.lean:188]:
```suggestion
/-- Division of an element by its radical in an Euclidean domain. -/
```

File: Mathlib/RingTheory/Radical.lean [hunk 2]
```diff
@@ -168,3 +179,51 @@ theorem radical_neg {a : R} : radical (-a) = radical a :=
...
+def divRadical (a : E) : E := a / radical a
+
+theorem mul_radical_divRadical (a : E) : radical a * divRadical a = a := by
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Radical.lean:191]: Infix notation also gets infixed in the name
```suggestion
theorem radical_mul_divRadical (a : E) : radical a * divRadical a = a := by
```

File: Mathlib/RingTheory/Radical.lean [hunk 3]
```diff
@@ -168,3 +179,51 @@ theorem radical_neg {a : R} : radical (-a) = radical a :=
...
+  exact right_ne_zero_of_mul ha
+
+theorem divRadical_isUnit {u : E} (hu : IsUnit u) : IsUnit (divRadical u) := by
+  rwa [divRadical, radical_unit_eq_one hu, EuclideanDomain.div_one]
+
+theorem eq_divRadical {a x : E} (h : radical a * x = a) : x = divRadical a := by
+  apply EuclideanDomain.eq_div_of_mul_eq_left (radical_ne_zero a)
+  rwa [mul_comm]
+
+theorem divRadical_mul {a b : E} (hc : IsCoprime a b) :
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Radical.lean:208]:
```suggestion
theorem divRadical_mul {a b : E} (hab : IsCoprime a b) :
```
is a bit clearer IMO

File: Mathlib/RingTheory/Radical.lean [hunk 4]
```diff
@@ -168,3 +179,51 @@ theorem radical_neg {a : R} : radical (-a) = radical a :=
...
+    divRadical (a * b) = divRadical a * divRadical b := by
+  by_cases ha : a = 0
+  · rw [ha, MulZeroClass.zero_mul, divRadical, EuclideanDomain.zero_div, MulZeroClass.zero_mul]
+  by_cases hb : b = 0
+  · rw [hb, MulZeroClass.mul_zero, divRadical, EuclideanDomain.zero_div, MulZeroClass.mul_zero]
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Radical.lean:213]: Looks like you're notactually usingthe extra assumotions?
```suggestion
```

File: Mathlib/RingTheory/Radical.lean [hunk 5]
```diff
@@ -168,3 +179,51 @@ theorem radical_neg {a : R} : radical (-a) = radical a :=
...
+  symm; apply eq_divRadical
+  rw [radical_mul hc]
+  rw [mul_mul_mul_comm, mul_radical_divRadical, mul_radical_divRadical]
+
+theorem divRadical_dvd_self (a : E) : divRadical a ∣ a := by
+  rw [divRadical]
+  apply EuclideanDomain.div_dvd_of_dvd
+  exact radical_dvd_self a
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Radical.lean:221]: Can you get there faster using mul_radical_divRadical ?

Comment (seewoo5) [reply Mathlib/RingTheory/Radical.lean]: Updated - I also added `divRadical_mul_radical` also (and used `Dvd.intro`).

File: Mathlib/RingTheory/Radical.lean [hunk 6]
```diff
@@ -168,3 +179,49 @@ theorem radical_neg {a : R} : radical (-a) = radical a :=
...
+theorem divRadical_mul {a b : E} (hab : IsCoprime a b) :
+    divRadical (a * b) = divRadical a * divRadical b := by
+  symm; apply eq_divRadical
+  rw [radical_mul hab]
+  rw [mul_mul_mul_comm, radical_mul_divRadical, radical_mul_divRadical]
+
+theorem divRadical_dvd_self (a : E) : divRadical a ∣ a := by
+  exact Dvd.intro (radical a) (divRadical_mul_radical a)
+
+theorem IsCoprime.divRadical {a b : E} (h : IsCoprime a b) :
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Radical.lean:221]:
```suggestion
protected theorem IsCoprime.divRadical {a b : E} (h : IsCoprime a b) :
```

**Review by YaelDillies — APPROVED** [review]
maintainer merge

apart from that. Thanks!

File: Mathlib/RingTheory/Radical.lean [hunk 7]
```diff
@@ -168,3 +179,49 @@ theorem radical_neg {a : R} : radical (-a) = radical a :=
...
+theorem divRadical_mul {a b : E} (hab : IsCoprime a b) :
+    divRadical (a * b) = divRadical a * divRadical b := by
+  symm; apply eq_divRadical
+  rw [radical_mul hab]
+  rw [mul_mul_mul_comm, radical_mul_divRadical, radical_mul_divRadical]
+
+theorem divRadical_dvd_self (a : E) : divRadical a ∣ a := by
+  exact Dvd.intro (radical a) (divRadical_mul_radical a)
+
+protected theorem IsCoprime.divRadical {a b : E} (h : IsCoprime a b) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Radical.lean:221]: What about calling this `_root_.IsCoprime.divRadical`?

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/RingTheory/Radical.lean (hunk 7 above)
Comment (seewoo5) [reply Mathlib/RingTheory/Radical.lean:221]: Honestly, I don't know which one would be better (`protected` vs `_root_`), could you elaborate more?

**Pull request closed without merging**